File: _stats_build/statistics/wr_variance.rb
require_relative "abstract/round_metric"

class WrVariance < RoundMetric
  def initialize
    @title = "Variance"
    @note = "Variance: sample variance of all 5 solves in a round (lower = more consistent)."
    @table_header = { "Result" => :right, "Improvement" => :right, "Days" => :right, "Person" => :left, "Competition" => :left, "Date" => :left, "Details" => :left }
  end

  # NOTE: Variance = 5次成绩的样本方差 (n-1)
  # 需要全部5次都有效
  def compute_metric(values, r)
    return nil unless values.all? { |v| v > 0 }
    mean = values.sum.to_f / values.length
    sum_sq = values.sum { |v| (v - mean) ** 2 }
    sum_sq / (values.length - 1)
  end

  # NOTE: 方差值用小数格式显示，不使用 SolveTime 格式化
  def format_metric(metric_value, event_id)
    # 将厘秒方差转换为秒方差（除以 10000）再显示
    variance_seconds = metric_value / 10000.0
    "%.3f" % variance_seconds
  end
end

File: core/packages/stats-build/src/statistics/wr_variance.ts
// NOTE: Variance——一轮中 5 次成绩的样本方差
// 与 Ruby _stats_build/statistics/wr_variance.rb 1:1 对应
import { RoundMetric } from '../core/round_metric.js';
import { EVENTS_WITH_AO5 } from '../core/events.js';

export class WrVariance extends RoundMetric {
  constructor() {
    super();
    this.title = 'Variance';
    this.titleZh = '方差';
    this.note = 'Variance: sample variance of all 5 solves in a round (lower = more consistent).';
    this.noteZh = '方差：一轮 5 次成绩的样本方差（越低越稳定）。';
    this.tableHeader = {
      'Result': 'right', 'Improvement': 'right', 'Days': 'right',
      'Person': 'left', 'Date': 'left', 'Competition': 'left', 'Details': 'left',
    };
  }

  targetEvents() { return EVENTS_WITH_AO5; }

  // NOTE: Variance = 样本方差 (n-1)，需要全部 5 次有效
  computeMetric(values: number[]): number | null {
    if (!values.every(v => v > 0)) return null;
    const mean = values.reduce((s, v) => s + v, 0) / values.length;
    const sumSq = values.reduce((s, v) => s + (v - mean) ** 2, 0);
    return sumSq / (values.length - 1);
  }

  // NOTE: 方差值显示为秒方差（厘秒² / 10000）
  formatMetric(metricValue: number): string {
    const varianceSeconds = metricValue / 10000.0;
    return varianceSeconds.toFixed(3);
  }
}
